rfcs/0136-scope-expression-registration.md

```diff
@@ -0,0 +1,129 @@
+   over-calling auth service. I think we can do this by making worker implementation part of the definition of a
+   workerType and having worker implementations generate the data that must be entered into the workerType definition.
+1. Build into the auth service a way to query all registered scope expression templates. The input will be a set of
+   scopes. The output will be a set of actions that may be authorized by these scopes and for each of these actions,
+   a list of possible values for parameters. This is a bit difficult to understand but the example should make this
+   more clear.
+
+# Example
+
+We create a new service called `taskcluster-dishwasher`. It washes dishes and has a single action called
+`dishwasher.wash`. This action is protected by a scope expression template:
+
+```json
+{
+  "AllOf": [
+    "dishwasher:wash:<detergent>"
+  ]
+}
+```
+
+The service will also define the term `detergent` with a bit of text about different values that might
+be present and also a regex that defines valid values for `detergent`.
+
+Before the service is started, when the cluster admin creates a client for this service, they will
+ensure that it has the scope `auth:register:dishwasher`. When the service is started, it will register
+this action and the associated terms with the auth service.
+
+Now the cluster admin would like to allow some users to use this service. They use either the tools site
+or ci-admin to add some scopes to roles. Before they do this however, they can use the built-in support
+in both of these tools for querying the auth service to see what the scopes they are giving out authorize.
+
+Because the auth service now has global knowledge of the scopes that control which actions, it can reliably
+tell you that for instance:
+
+* Giving a client the scope `dishwasher:*` allows them to call `dishwasher.wash` with _any_ detergent.
+* Giving a client the scope `dishwasher:wash:ajax-*` allows them to call `dishwasher.wash` with any
+  detergent that begins with `ajax-`.
+* Given a client the scope `dishwasher:wash:comet` allows them to call `dishwasher.wash` but only with the
+  detergent `comet`.
+* If a client does not have any `dishwasher` scopes, we know that they cannot do anything with the
+  dishwashing service.
+
+This would work similarly for workers. It can even be helpful if we just register principals as owning
+a specific scope namespace as we can tell roughly what a given set of scopes permits -- although with
+significantly less granularity.
+
+# Open Questions
+
+Many of the finer points of this are quite hand-wavy. We can try to button things down a bit before
+moving forward if we like.
+
+This does require auth to be running for any other service to start. Maybe we can be smart by retrying
+registration until auth is running or just always checking to see if auth has the most recent version
+of the definition every time we call `authorize()`?
```

Comment:
I think it's OK for a service to not really be "up" until auth is started -- after all, it's unlikely to work correctly without auth anyway.  Running in Kubernetes means that the existing strategy of exiting the process when startup fails is a good fix, since k8s will automatically restart the process -- so when auth finally comes up, the dependent service will too (Heroku isn't so nice..)